compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -464,11 +467,143 @@ fn check_opaque_meets_bounds<'tcx>(
             ocx.resolve_regions_and_report_errors(defining_use_anchor, &outlives_env)?;
         }
     }
-    // Clean up after ourselves
-    let _ = infcx.take_opaque_types();
+    // Check that any hidden types found during wf checking match the hidden types that `type_of` sees.
+    for (key, mut ty) in infcx.take_opaque_types() {
+        ty.hidden_type.ty = infcx.resolve_vars_if_possible(ty.hidden_type.ty);
+        sanity_check_found_hidden_type(tcx, key, ty.hidden_type, defining_use_anchor, origin)?;
+    }
     Ok(())
 }
 
+fn sanity_check_found_hidden_type<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    key: ty::OpaqueTypeKey<'tcx>,
+    mut ty: ty::OpaqueHiddenType<'tcx>,
+    defining_use_anchor: LocalDefId,
+    origin: &hir::OpaqueTyOrigin,
+) -> Result<(), ErrorGuaranteed> {
+    if ty.ty.is_ty_var() {
+        // Nothing was actually constrained.
+        return Ok(());
+    }
+    if let ty::Alias(ty::Opaque, alias) = ty.ty.kind() {
+        if alias.def_id == key.def_id.to_def_id() && alias.args == key.args {
+            // Nothing was actually constrained, this is an opaque usage that was
+            // only discovered to be opaque after inference vars resolved.
+            return Ok(());
+        }
+    }
+    // Closures frequently end up containing erased lifetimes in their final representation.
+    // These correspond to lifetime variables that never got resolved, so we patch this up here.
+    ty.ty = ty.ty.fold_with(&mut BottomUpFolder {
+        tcx,
+        ty_op: |t| t,
+        ct_op: |c| c,
+        lt_op: |l| match l.kind() {
+            RegionKind::ReVar(_) => tcx.lifetimes.re_erased,
+            _ => l,
+        },
+    });
+    // Get the hidden type, and in case it is in a nested opaque type, find that opaque type's
+    // usage in the function signature and use the generic arguments from the usage site.
+    let mut hidden_ty = tcx.type_of(key.def_id).instantiate(tcx, key.args);
+    if let hir::OpaqueTyOrigin::FnReturn(..) | hir::OpaqueTyOrigin::AsyncFn(..) = origin {
+        if hidden_ty != ty.ty {
+            hidden_ty = find_and_apply_rpit_substs(
+                tcx,
+                hidden_ty,
+                defining_use_anchor.to_def_id(),
+                key.def_id.to_def_id(),
+            )?;
+        }
```

Comment:
I am confused by this :thinking: 

so we're checking some opaque type `fn foo<'a T>() -> impl Foo + 'a` which defines some other opaque `impl Trait<T>`

if `impl Foo` is an RPIT and `type_of(impl Trait<T>).subst(T)`, we call `find_and_apply_rpit_substs` for `impl Trait<T>`. What exactly happens in that function? :thinking: it seems like it always uses a `delay_span_bug`?

do you have a concrete example where `find_and_apply_rpit_substs` is necessary?

Reply:
I guess the comment

> // Get the hidden type, and in case it is in a nested opaque type, find that opaque type's
>    // usage in the function signature and use the generic arguments from the usage site.
 
doesn't really make snese to me :thinking:

Reply:
```
error: concrete type differs from previous defining opaque type use
  --> /home/ubuntu/rustc/rust2/tests/ui/type-alias-impl-trait/nested-rpit-with-lifetimes.rs:42:84
   |
LL | pub fn metric_stream_parser<'a, I>() -> impl Parser<I, Output = (), PartialState = impl Any + 'a>
   |                                                                                    ^^^^^^^^^^^^^
   |                                                                                    |
   |                                                                                    expected `Map<&ReEarlyBound(DefId(0:39 ~ nested_rpit_with_lifetimes[8f41]::metric_stream_parser::'a), 0, 'a) [()]>`, got `Map<&ReEarlyBound(DefId(0:43 ~ nested_rpit_with_lifetimes[8f41]::metric_stream_parser::{opaque#0}::'a), 2, 'a) [()]>`
   |                                                                                    this expression supplies two conflicting concrete types for the same opaque type
```

from https://github.com/rust-lang/rust/blob/2d0b97c64cdc0eaefbd76f16fb75930c90666e8c/tests/ui/type-alias-impl-trait/nested-rpit-with-lifetimes.rs

Reply:
This all happens during the check_opaque_meets_bounds of `DefId(0:42 ~ nested_rpit_with_lifetimes[8f41]::metric_stream_parser::{opaque#0})` (so the outer RPIT).

`type_of` of the nested opaque type correctly returns

```
Map<&ReEarlyBound(DefId(0:45 ~ nested_rpit_with_lifetimes[8f41]::metric_stream_parser::{opaque#0}::{opaque#0}::'a), 2, 'a) [()]>
```

but this gets then substituted with

```
[
    ReEarlyBound(DefId(0:39 ~ nested_rpit_with_lifetimes[8f41]::metric_stream_parser::'a), 0, 'a),
    I,
    ReEarlyBound(DefId(0:43 ~ nested_rpit_with_lifetimes[8f41]::metric_stream_parser::{opaque#0}::'a), 2, 'a)]
```

which is why we end up with

```
Map<&ReEarlyBound(DefId(0:45 ~ nested_rpit_with_lifetimes[8f41]::metric_stream_parser::{opaque#0}::{opaque#0}::'a), 2, 'a) [()]>
```

instead of the version that is described solely in terms of the function's generics.

Reply:
The issue is that opaque types ignore the lifetimes of the function, as they have their own copies. So the only way to get the lifetimes represented in terms of the function, is to look how they are used in the function signature (or do some other fancy recording of this mapping at hir lowering time).

Ideally we'd not have this lifetime duplication and dance, but that's a separate story.